### PR TITLE
Add a server endpoint for serving devtools extensions

### PR DESCRIPTION
**Pull request by kenzieschmoll**
This PR
- moves `extension_model.dart` to `packages/devtools_shared` so that we can reuse this model from the devtools server
- creates `ExtensionManager` that manages moving devtools extension assets to `build/devtools_extensions` so that the server can serve them (implemented in https://dart-review.googlesource.com/c/sdk/+/305043)
- adds a server endpoint to serve devtools extensions

work towards https://github.com/flutter/devtools/issues/1632

### CODE REVIEW

File: packages/devtools_app/lib/src/extensions/extension_screen.dart [hunk 1]
```diff
@@ -155,3 +155,10 @@ class EmbeddedExtensionHeader extends StatelessWidget {
     );
   }
 }
+
+extension ExtensionConfigExtension on DevToolsExtensionConfig {
+  IconData get icon => IconData(
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/extensions/extension_screen.dart:160]: Why does this need to be an extension instead of being a property of `DevToolsExtensionConfig`?

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+/// The location for the DevTools extension, relative to the parent package's
+/// root.
+const extensionLocation = 'extension/devtools';
+
+/// The default location for the DevTools extension, relative to
+/// `<parent_package_root>/extension/devtools/`.
+const extensionBuildDefault = 'build';
+
+class ExtensionsManager {
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_manager.dart:24]: What does this class do? We should have some documentation here.

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  ExtensionsManager({required this.buildDir});
+
+  final String buildDir;
+
+  String get servedExtensionsPath => path.join(buildDir, extensionRequestPath);
+
+  final devtoolsExtensions = <DevToolsExtensionConfig>[];
+
+  void serveAvailableExtensions(String? rootPath) {
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_manager.dart:33]: This needs documentation

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    String extensionPackageName,
+    String extensionPath,
+  ) {
+    final newExtensionPath =
+        path.join(servedExtensionsPath, extensionPackageName);
+    copyPathSync(extensionPath, newExtensionPath);
+  }
+}
+
+/// Mark copied from package:extension_discovery. Remove once it is published.
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_manager.dart:113]: Is this comment related to `Extension`? Maybe this should be a TODO?

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+/// Information about an extension for target package.
+class Extension {
+  Extension._({
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_manager.dart:117]: Does anything create an instance of `Extension` yet? I don't see any references in this library and it doesn't have a public constructor.

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    required this.config,
+  });
+
+  /// Name of the package providing an extension.
+  final String package;
+
+  /// Absolute path to the package root.
+  ///
+  /// This folder usually contains a `lib/` folder and a `pubspec.yaml`
+  /// (assuming dependencies are fetched using the pub package manager).
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_manager.dart:130]: We make references to fetching using `dart pub` below, but use "fetched using the pub package manager" verbiage here. We should probably pick one or the other (FWIW, I prefer just saying "dart pub").

File: packages/devtools_shared/lib/src/extensions/extension_model.dart [hunk 7]
```diff
@@ -0,0 +1,74 @@
+// Copyright 2023 The Chromium Authors. All rights reserved.
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_model.dart:1]: Just confirming, this is a straight move from `package:devtools_app/src/extensions/embedded/extension_model.dart` right?

**Review by bkonyi — COMMENTED** [review]
First pass.

File: packages/devtools_app/lib/src/shared/config_specific/server/_server_web.dart [hunk 8]
```diff
@@ -337,6 +338,31 @@ DevToolsJsonFile _devToolsJsonFileFromResponse(
...
+    final uri = Uri(
+      path: apiServeAvailableExtensions,
+      queryParameters: {extensionRootPathPropertyName: rootPath},
+    );
+    final resp = await request(uri.toString());
+    if (resp?.status == HttpStatus.ok) {
+      final parsedResult = json.decode(resp!.responseText!);
+      final extensionsAsJson =
+          (parsedResult[extensionsResultPropertyName]! as List<Object?>)
+              .cast<Map<String, Object?>>();
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/config_specific/server/_server_web.dart:354]: There's a possibility this cast could fail. If the result is `List<Object?>`, casting to a non-nullable `List<Map<String, Object?>>` would throw if there is a null entry. It probably doesn't make much of a difference since we'd throw during the `as List<Object>` cast if there's a null entry in the list, but can we instead do `(result as List<Object>).cast<Map<String, Object?>>()` to make it clear that we don't expect a null entry?

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+const extensionLocation = 'extension/devtools';
+
+/// The default location for the DevTools extension, relative to
+/// `<parent_package_root>/extension/devtools/`.
+const extensionBuildDefault = 'build';
+
+class ExtensionsManager {
+  ExtensionsManager({required this.buildDir});
+
+  final String buildDir;
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_manager.dart:27]: All public properties and methods on this class need documentation.

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+      // TODO(kenz): use 'findExtensions' from package:extension_discovery once it
+      // is published.
+      // final extensions = findExtensions(
+      //   'devtools',
+      //   packageConfig: '$rootPath/.dart_tool/package_config.json',
+      // );
+      final extensions = <Extension>[];
+      for (final extension in extensions) {
+        final config = extension.config;
+        if (config is! Map) {
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_manager.dart:46]: If `config` is always expected to be `Map`, why don't we restrict it to `Map<Object, Object?>` in `Extension`?

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+    _resetServedPluginsDir();
+    for (final extension in devtoolsExtensions) {
+      _moveToServedExtensionsDir(extension.name, extension.path);
+    }
+  }
+
+  void _resetServedPluginsDir() {
+    final buildDirectory = Directory(buildDir);
+    if (!buildDirectory.existsSync()) {
+      throw Exception('The build directory does not exist.');
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_manager.dart:91]: Can we use a more specific exception type here?

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    servedExtensionsDir.createSync();
+  }
+
+  void _moveToServedExtensionsDir(
+    String extensionPackageName,
+    String extensionPath,
+  ) {
+    final newExtensionPath =
+        path.join(servedExtensionsPath, extensionPackageName);
+    copyPathSync(extensionPath, newExtensionPath);
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_manager.dart:109]: In general, we should avoid using `fooSync()` APIs in production as they block the event loop for the duration of the call.

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    }
+    servedExtensionsDir.createSync();
+  }
+
+  void _moveToServedExtensionsDir(
+    String extensionPackageName,
+    String extensionPath,
+  ) {
+    final newExtensionPath =
+        path.join(servedExtensionsPath, extensionPackageName);
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_manager.dart:108]: Style nit: consider adding a trailing comma

**Review by bkonyi — COMMENTED** [review]
Another pass.

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart (hunk 10 above)
Comment (kenzieschmoll) [reply packages/devtools_shared/lib/src/extensions/extension_manager.dart:46]: See discussion at https://github.com/dart-lang/tools/pull/129#discussion_r1269800754

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart (hunk 4 above)
Comment (kenzieschmoll) [reply packages/devtools_shared/lib/src/extensions/extension_manager.dart:113]: Yes everything below here will be published from package:extension_discovery, but I  copied the type here so I can implement `serveAvailableExtensions` using this type.

Initial PR for package:extension_discovery is here: https://github.com/dart-lang/tools/pull/129

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart (hunk 5 above)
Comment (kenzieschmoll) [reply packages/devtools_shared/lib/src/extensions/extension_manager.dart:117]: We use the type on line 43.

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart (hunk 6 above)
Comment (kenzieschmoll) [reply packages/devtools_shared/lib/src/extensions/extension_manager.dart:130]: Ack. But this code is copied from https://github.com/dart-lang/tools/pull/129, and it will eventually be removed from devtools_shared. It is only in this package temporarily. I will prune the comments and make it a private class to make that more clear.

File: packages/devtools_app/lib/src/extensions/extension_screen.dart (hunk 1 above)
Comment (bkonyi) [reply packages/devtools_app/lib/src/extensions/extension_screen.dart:160]: Makes sense, thanks for explaining!

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+    devtoolsExtensions.clear();
+
+    if (rootPath != null) {
+      // TODO(kenz): use 'findExtensions' from package:extension_discovery once it
+      // is published.
+      // final extensions = findExtensions(
+      //   'devtools',
+      //   packageConfig: '$rootPath/.dart_tool/package_config.json',
+      // );
+      final extensions = <Extension>[];
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_manager.dart:43]: I think this is an error now that you've renamed `Extension` to `_Extension`.

File: packages/devtools_shared/lib/src/devtools_api.dart [hunk 15]
```diff
@@ -64,3 +64,12 @@ const String apiGetTestAppSizeFile = '${apiPrefix}getTestAppSizeFile';
 const String baseAppSizeFilePropertyName = 'appSizeBase';
 
 const String testAppSizeFilePropertyName = 'appSizeTest';
+
+/// Serves any available extensions and returns a list of their configurations
+/// to DevTools.
+const String apiServeAvailableExtensions =
+    '${apiPrefix}serveAvailableExtensions';
+
+const extensionRootPathPropertyName = 'rootPath';
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/devtools_api.dart:73]: Nit: documentation for these two property strings. Maybe add types to be consistent with the rest of the constants here?

File: packages/devtools_shared/lib/src/extensions/extension_model.dart (hunk 7 above)
Comment (kenzieschmoll) [reply packages/devtools_shared/lib/src/extensions/extension_model.dart:1]: I added dart doc to the members and some validation in the factory constructor so it should be re-reviewed. I also made the json parsing for the icon code point able to handle both strings and ints

File: packages/devtools_app/lib/src/shared/config_specific/server/_server_web.dart (hunk 8 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/shared/config_specific/server/_server_web.dart:354]: Replacing `List<Object?>` with `List<Object>` resulted in this error:
```
TypeError: Instance of 'JSArray<dynamic>': type 'JSArray<dynamic>' is not a subtype of type 'List<Object>'
```

Instead, I added a `.whereNotNull()` call in the chain to ensure we aren't casting a potentially null element to a non null map.

File: packages/devtools_shared/lib/src/extensions/extension_manager.dart (hunk 11 above)
Comment (kenzieschmoll) [reply packages/devtools_shared/lib/src/extensions/extension_manager.dart]: used FileSystemException

File: packages/devtools_app/lib/src/shared/config_specific/server/_server_web.dart (hunk 8 above)
Comment (bkonyi) [reply packages/devtools_app/lib/src/shared/config_specific/server/_server_web.dart:356]: Huh, that's interesting... Using `.whereNotNull()` is a good call.

File: packages/devtools_shared/lib/src/extensions/extension_model.dart [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    final codePointFromJson = json[materialIconCodePointKey];
+    const defaultCodePoint = 0xf03f;
+    if (codePointFromJson is String?) {
+      codePoint =
+          int.tryParse(codePointFromJson ?? '0xf03f') ?? defaultCodePoint;
+    } else {
+      codePoint = codePointFromJson as int? ?? defaultCodePoint;
+    }
+
+    final name = json[nameKey] as String?;
```
Comment (bkonyi) [inline packages/devtools_shared/lib/src/extensions/extension_model.dart:28]: We should use pattern matching here for JSON validation:

```dart
  if (json
      case {
        nameKey: String name,
        pathKey: String path,
        issueTrackerKey: String issueTracker,
        versionKey: String version,
      }) {
    return DevToolsExtensionConfig._(
      name: name,
      path: path,
      issueTrackerLink: issueTrackerLink,
      version: version,
      materialIconCodePoint: codePoint,
    );
  }
  throw StateError(
    'missing required fields ${nullFields.toString()} in the extension '
    'config.json',
  );
```

**Review by bkonyi — APPROVED** [review]
LGTM with one last suggestion.

File: packages/devtools_shared/lib/src/extensions/extension_model.dart (hunk 16 above)
Comment (kenzieschmoll) [reply packages/devtools_shared/lib/src/extensions/extension_model.dart:28]: I will address this comment as part of https://github.com/flutter/devtools/pull/6097.
